imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -368,6 +382,101 @@ def set_spin_cycle(
     return l1b_de
 
 
+# TODO: The spin cycle function above needs to be updated for DEs. We cannot assume
+#  there are 28 spins per ASC and we should calculate the spin start number based on the
+#  corresponding L1A spin data Acq Start for the ASC. The implementation below should be
+#  should be used for the DE rather than the above function, but in the interest of time
+#  the below function is only hooked up to the histogram rates processing and should be
+#  integrated into the DE processing in a later PR.
+# TODO: Break up the invalid spin ASC removal and the code to find the closest DE/Hist
+#  and spin ASCs into their own functions.
+def set_spin_cycle_from_spin_data(
+    l1a_science: xr.Dataset, l1b_science: xr.Dataset, spin_data: xr.Dataset
+) -> xr.Dataset:
+    """
+    Set the spin cycle for each direct event using the L1A spin data.
+
+    Parameters
+    ----------
+    l1a_science : xr.Dataset
+        The L1A Histogram or Direct Event dataset.
+    l1b_science : xr.Dataset
+        The L1B Histogram Rate or Direct Event dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+
+    Returns
+    -------
+    l1b_science : xr.Dataset
+        The L1B science dataset with the spin cycle added for each direct event.
+    """
+    acq_start, _acq_end = convert_start_end_acq_times(spin_data)
+
+    spin_met_per_asc = spin_data["shcoarse"].values
+    science_met_per_asc = ttj2000ns_to_met(l1a_science["epoch"])
+
+    # Find the closest start_acq for each direct event
+    # computes the index of the closest spin_met_per_asc for each science_met_per_asc
+    # so the resulting array will be of length len(science_met_per_asc), one index per
+    # ASC, but the value of each index will be the index of the closest spin data.
+    closest_start_acq_indices = np.abs(
+        science_met_per_asc[:, None] - spin_met_per_asc
+    ).argmin(axis=1)
+
+    valid_asc_mask = []
+    for spin_idx in closest_start_acq_indices:
+        spin_count = spin_data["num_completed"].isel(epoch=spin_idx)
+        valid_asc_mask.append(spin_count >= 28)
+
+    valid_asc_mask = np.array(valid_asc_mask)
+
+    # Filter out ASCs with insufficient spins
+    if not valid_asc_mask.all():
+        logger.warning(
+            f"Dropping {(~valid_asc_mask).sum()} ASCs with fewer than 28 spins"
+        )
+        l1a_science = l1a_science.isel(epoch=valid_asc_mask)
+        l1b_science = l1b_science.isel(epoch=valid_asc_mask)
+        science_met_per_asc = science_met_per_asc[valid_asc_mask]
+        closest_start_acq_indices = closest_start_acq_indices[valid_asc_mask]
+
+    closest_start_acq_per_asc = acq_start.isel(
+        epoch=xr.DataArray(closest_start_acq_indices)
+    )
+    # Get the spin cycle number from the spin data for each direct event
+    spin_start_num_per_asc = np.atleast_1d(get_spin_number(closest_start_acq_per_asc))
```

Comment:
Could you put in a comment about "spin_cycle" being the average spin number for that ESA step during that ASC? It was not clear to me after our discussion on Monday that the cycle was an average value because we repeat the ESA step pattern twice with 0-13 containing the first sequence and 14-27 containing the second sequence so we are averaging over those two sequences.